File: LocalPayment/src/test/java/com/braintreepayments/api/localpayment/LocalPaymentApiUnitTest.java
package com.braintreepayments.api.localpayment;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertSame;
import static org.junit.Assert.assertTrue;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyMap;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.ArgumentMatchers.isNull;
import static org.mockito.ArgumentMatchers.same;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import android.net.Uri;

import com.braintreepayments.api.core.AnalyticsParamRepository;
import com.braintreepayments.api.core.BraintreeClient;
import com.braintreepayments.api.core.IntegrationType;
import com.braintreepayments.api.core.PostalAddress;
import com.braintreepayments.api.sharedutils.HttpResponseCallback;
import com.braintreepayments.api.testutils.Fixtures;
import com.braintreepayments.api.testutils.MockBraintreeClientBuilder;

import org.json.JSONException;
import org.json.JSONObject;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.robolectric.RobolectricTestRunner;
import org.skyscreamer.jsonassert.JSONAssert;

@RunWith(RobolectricTestRunner.class)
public class LocalPaymentApiUnitTest {

    private LocalPaymentInternalAuthRequestCallback localPaymentInternalAuthRequestCallback;
    private LocalPaymentInternalTokenizeCallback localPaymentInternalTokenizeCallback;

    private AnalyticsParamRepository analyticsParamRepository;

    @Before
    public void beforeEach() {
        analyticsParamRepository = mock(AnalyticsParamRepository.class);
        localPaymentInternalAuthRequestCallback =
                mock(LocalPaymentInternalAuthRequestCallback.class);
        localPaymentInternalTokenizeCallback =
                mock(LocalPaymentInternalTokenizeCallback.class);

        when(analyticsParamRepository.getSessionId()).thenReturn("sample-session-id");
    }

    @Test
    public void createPaymentMethod_sendsCorrectPostParams() throws JSONException {
        BraintreeClient braintreeClient = new MockBraintreeClientBuilder()
                .returnUrlScheme("sample-scheme")
                .build();

        LocalPaymentApi sut = new LocalPaymentApi(braintreeClient, analyticsParamRepository);
        sut.createPaymentMethod(
                getIdealLocalPaymentRequest(),
                localPaymentInternalAuthRequestCallback
        );

        String expectedPath = "/v1/local_payments/create";
        ArgumentCaptor<String> bodyCaptor = ArgumentCaptor.forClass(String.class);
        verify(braintreeClient).sendPOST(
                eq(expectedPath), bodyCaptor.capture(), anyMap(),
                any(HttpResponseCallback.class));

        String requestBody = bodyCaptor.getValue();
        JSONObject json = new JSONObject(requestBody);
        assertEquals("Doe", json.getString("lastName"));
        assertEquals("1.10", json.getString("amount"));
        assertEquals("Den Haag", json.getString("city"));
        assertEquals("2585 GJ", json.getString("postalCode"));
        assertEquals("sale", json.getString("intent"));
        assertEquals("Jon", json.getString("firstName"));
        assertEquals("639847934", json.getString("phone"));
        assertEquals("NL", json.getString("countryCode"));
        assertEquals("EUR", json.getString("currencyIsoCode"));
        assertEquals("ideal", json.getString("fundingSource"));
        assertEquals("jon@getbraintree.com", json.getString("payerEmail"));
        assertEquals("836486 of 22321 Park Lake", json.getString("line1"));
        assertEquals("Apt 2", json.getString("line2"));
        assertEquals("CA", json.getString("state"));
        assertEquals("local-merchant-account-id", json.getString("merchantAccountId"));
        assertTrue(json.getJSONObject("experienceProfile").getBoolean("noShipping"));
        assertEquals(
                "My Brand!",
                json.getJSONObject("experienceProfile").getString("brandName")
        );
        String expectedCancelUrl = Uri.parse("sample-scheme://local-payment-cancel").toString();
        String expectedReturnUrl = Uri.parse("sample-scheme://local-payment-success").toString();
        assertEquals(expectedCancelUrl, json.getString("cancelUrl"));
        assertEquals(expectedReturnUrl, json.getString("returnUrl"));
    }

    @Test
    public void createPaymentMethod_onPOSTError_forwardsErrorToCallback() {
        Exception error = new Exception("error");
        BraintreeClient braintreeClient = new MockBraintreeClientBuilder()
                .sendPOSTErrorResponse(error)
                .build();

        LocalPaymentApi sut = new LocalPaymentApi(braintreeClient, analyticsParamRepository);
        sut.createPaymentMethod(
                getIdealLocalPaymentRequest(),
                localPaymentInternalAuthRequestCallback
        );

        verify(localPaymentInternalAuthRequestCallback).onLocalPaymentInternalAuthResult(
                isNull(),
                same(error)
        );

    }

    @Test
    public void createPaymentMethod_onJSONError_forwardsJSONErrorToCallback() {
        BraintreeClient braintreeClient = new MockBraintreeClientBuilder()
                .sendPOSTSuccessfulResponse(Fixtures.ERROR_RESPONSE)
                .build();

        LocalPaymentApi sut = new LocalPaymentApi(braintreeClient, analyticsParamRepository);
        sut.createPaymentMethod(
                getIdealLocalPaymentRequest(),
                localPaymentInternalAuthRequestCallback
        );

        ArgumentCaptor<Exception> captor = ArgumentCaptor.forClass(Exception.class);
        verify(localPaymentInternalAuthRequestCallback).onLocalPaymentInternalAuthResult(
                isNull(),
                captor.capture()
        );

        assertTrue(captor.getValue() instanceof JSONException);
    }

    @Test
    public void createPaymentMethod_onPOSTSuccess_returnsResultWithOriginalRequestToCallback() {
        BraintreeClient braintreeClient = new MockBraintreeClientBuilder()
                .sendPOSTSuccessfulResponse(Fixtures.PAYMENT_METHODS_LOCAL_PAYMENT_CREATE_RESPONSE)
                .build();

        LocalPaymentApi sut = new LocalPaymentApi(braintreeClient, analyticsParamRepository);
        LocalPaymentRequest request = getIdealLocalPaymentRequest();
        sut.createPaymentMethod(request, localPaymentInternalAuthRequestCallback);

        ArgumentCaptor<LocalPaymentAuthRequestParams> captor =
                ArgumentCaptor.forClass(LocalPaymentAuthRequestParams.class);
        verify(localPaymentInternalAuthRequestCallback).onLocalPaymentInternalAuthResult(
                captor.capture(),
                isNull()
        );

        LocalPaymentAuthRequestParams result = captor.getValue();
        assertNotNull(result);
        assertSame(request, result.getRequest());
        assertEquals(
                "https://checkout.paypal.com/latinum?token=payment-token",
                result.getApprovalUrl()
        );
        assertEquals("local-payment-id-123", result.getPaymentId());
    }

    @Test
    public void tokenize_sendsCorrectPostParams() throws JSONException {
        BraintreeClient braintreeClient = new MockBraintreeClientBuilder()
                .integration(IntegrationType.CUSTOM)
                .build();

        LocalPaymentApi sut = new LocalPaymentApi(braintreeClient, analyticsParamRepository);
        String webUrl = "sample-scheme://local-payment-success?paymentToken=successTokenId";
        sut.tokenize(
                "local-merchant-account-id", webUrl, "sample-correlation-id",
                localPaymentInternalTokenizeCallback
        );

        ArgumentCaptor<String> bodyCaptor = ArgumentCaptor.forClass(String.class);
        String expectedUrl = "/v1/payment_methods/paypal_accounts";

        verify(braintreeClient).sendPOST(
                eq(expectedUrl), bodyCaptor.capture(),anyMap(),
                any(HttpResponseCallback.class));
        String requestBody = bodyCaptor.getValue();

        JSONObject expectedJSON = new JSONObject();
        expectedJSON.put("merchant_account_id", "local-merchant-account-id");

        JSONObject paypalAccount = new JSONObject()
                .put("intent", "sale")
                .put("response", new JSONObject().put("webURL", webUrl))
                .put("options", new JSONObject().put("validate", false))
                .put("response_type", "web")
                .put("correlation_id", "sample-correlation-id");
        expectedJSON.put("paypal_account", paypalAccount);

        JSONObject metaData = new JSONObject()
                .put("source", "client")
                .put("integration", "custom")
                .put("sessionId", "sample-session-id");
        expectedJSON.put("_meta", metaData);

        JSONAssert.assertEquals(expectedJSON, new JSONObject(requestBody), true);
    }

    @Test
    public void tokenize_onPOSTError_forwardsErrorToCallback() {
        Exception error = new Exception("error");
        BraintreeClient braintreeClient = new MockBraintreeClientBuilder()
                .integration(IntegrationType.CUSTOM)
                .sendPOSTErrorResponse(error)
                .build();

        LocalPaymentApi sut = new LocalPaymentApi(braintreeClient, analyticsParamRepository);
        String webUrl = "sample-scheme://local-payment-success?paymentToken=successTokenId";
        sut.tokenize(
                "local-merchant-account-id", webUrl, "sample-correlation-id",
                localPaymentInternalTokenizeCallback
        );

        verify(localPaymentInternalTokenizeCallback).onResult(isNull(), same(error));
    }

    @Test
    public void tokenize_onJSONError_forwardsErrorToCallback() {
        BraintreeClient braintreeClient = new MockBraintreeClientBuilder()
                .integration(IntegrationType.CUSTOM)
                .sendPOSTSuccessfulResponse("not-json")
                .build();

        LocalPaymentApi sut = new LocalPaymentApi(braintreeClient, analyticsParamRepository);
        String webUrl = "sample-scheme://local-payment-success?paymentToken=successTokenId";
        sut.tokenize(
                "local-merchant-account-id", webUrl, "sample-correlation-id",
                localPaymentInternalTokenizeCallback
        );

        ArgumentCaptor<Exception> captor = ArgumentCaptor.forClass(Exception.class);
        verify(localPaymentInternalTokenizeCallback).onResult(isNull(), captor.capture());

        assertTrue(captor.getValue() instanceof JSONException);
    }

    @Test
    public void tokenize_onPOSTSuccess_returnsResultToCallback() {
        BraintreeClient braintreeClient = new MockBraintreeClientBuilder()
                .integration(IntegrationType.CUSTOM)
                .sendPOSTSuccessfulResponse(Fixtures.PAYMENT_METHODS_LOCAL_PAYMENT_RESPONSE)
                .build();

        LocalPaymentApi sut = new LocalPaymentApi(braintreeClient, analyticsParamRepository);
        String webUrl = "sample-scheme://local-payment-success?paymentToken=successTokenId";
        sut.tokenize(
                "local-merchant-account-id", webUrl, "sample-correlation-id",
                localPaymentInternalTokenizeCallback
        );

        ArgumentCaptor<LocalPaymentNonce> captor = ArgumentCaptor.forClass(LocalPaymentNonce.class);
        verify(localPaymentInternalTokenizeCallback).onResult(
                captor.capture(),
                (Exception) isNull()
        );

        LocalPaymentNonce result = captor.getValue();
        assertNotNull(result);
        assertEquals("e11c9c39-d6a4-0305-791d-bfe680ef2d5d", result.getString());
        assertEquals("084afbf1db15445587d30bc120a23b09", result.getClientMetadataId());
        assertEquals("jon@getbraintree.com", result.getEmail());
        assertEquals("Jon", result.getGivenName());
        assertEquals("Doe", result.getSurname());
        assertEquals("9KQSUZTL7YZQ4", result.getPayerId());

        PostalAddress shippingAddress = result.getShippingAddress();
        assertEquals("Jon Doe", shippingAddress.getRecipientName());
        assertEquals("836486 of 22321 Park Lake", shippingAddress.getStreetAddress());
    }

    private LocalPaymentRequest getIdealLocalPaymentRequest() {
        PostalAddress address = new PostalAddress();
        address.setStreetAddress("836486 of 22321 Park Lake");
        address.setExtendedAddress("Apt 2");
        address.setCountryCodeAlpha2("NL");
        address.setLocality("Den Haag");
        address.setRegion("CA");
        address.setPostalCode("2585 GJ");

        LocalPaymentRequest request = new LocalPaymentRequest(true);
        request.setPaymentType("ideal");
        request.setAmount("1.10");
        request.setAddress(address);
        request.setPhone("639847934");
        request.setEmail("jon@getbraintree.com");
        request.setGivenName("Jon");
        request.setSurname("Doe");
        request.setShippingAddressRequired(false);
        request.setMerchantAccountId("local-merchant-account-id");
        request.setCurrencyCode("EUR");
        request.setPaymentTypeCountryCode("NL");
        request.setDisplayName("My Brand!");

        return request;
    }
}

File: LocalPayment/src/test/java/com/braintreepayments/api/localpayment/LocalPaymentApiUnitTest.kt
package com.braintreepayments.api.localpayment

import android.net.Uri
import com.braintreepayments.api.core.AnalyticsParamRepository
import com.braintreepayments.api.core.BraintreeException
import com.braintreepayments.api.core.IntegrationType
import com.braintreepayments.api.core.MerchantRepository
import com.braintreepayments.api.core.PostalAddress
import com.braintreepayments.api.testutils.Fixtures
import com.braintreepayments.api.testutils.MockkBraintreeClientBuilder
import io.mockk.every
import io.mockk.mockk
import io.mockk.slot
import io.mockk.verify
import org.json.JSONException
import org.json.JSONObject
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.skyscreamer.jsonassert.JSONAssert

@RunWith(RobolectricTestRunner::class)
class LocalPaymentApiUnitTest {
    private lateinit var localPaymentInternalAuthRequestCallback: LocalPaymentInternalAuthRequestCallback
    private lateinit var localPaymentInternalTokenizeCallback: LocalPaymentInternalTokenizeCallback

    private lateinit var analyticsParamRepository: AnalyticsParamRepository
    private lateinit var merchantRepository: MerchantRepository

    @Before
    fun beforeEach() {
        analyticsParamRepository = mockk(relaxed = true)
        localPaymentInternalAuthRequestCallback = mockk(relaxed = true)
        localPaymentInternalTokenizeCallback = mockk(relaxed = true)
        merchantRepository = mockk(relaxed = true)

        every { analyticsParamRepository.sessionId } returns "sample-session-id"
        every { merchantRepository.integrationType } returns IntegrationType.CUSTOM
    }

    @Test
    fun createPaymentMethod_sendsCorrectPostParams() {
        val braintreeClient = MockkBraintreeClientBuilder()
            .returnUrlScheme("sample-scheme")
            .build()

        val sut = LocalPaymentApi(braintreeClient, analyticsParamRepository, merchantRepository)
        sut.createPaymentMethod(
            getIdealLocalPaymentRequest(),
            localPaymentInternalAuthRequestCallback
        )

        val captor = slot<String>()
        verify { braintreeClient.sendPOST(eq("/v1/local_payments/create"), capture(captor), any(), any()) }
        val requestBody = captor.captured

        val expectedJSON = JSONObject()
        expectedJSON.put("intent", "sale")
        expectedJSON.put("returnUrl", "sample-scheme://local-payment-success")
        expectedJSON.put("cancelUrl", "sample-scheme://local-payment-cancel")
        expectedJSON.put("fundingSource", "ideal")
        expectedJSON.put("amount", "1.10")
        expectedJSON.put("currencyIsoCode", "EUR")
        expectedJSON.put("firstName", "Jon")
        expectedJSON.put("lastName", "Doe")
        expectedJSON.put("payerEmail", "jon@getbraintree.com")
        expectedJSON.put("phone", "639847934")
        expectedJSON.put("merchantAccountId", "local-merchant-account-id")
        expectedJSON.put("paymentTypeCountryCode", "NL")
        expectedJSON.put("line1", "836486 of 22321 Park Lake")
        expectedJSON.put("line2", "Apt 2")
        expectedJSON.put("city", "Den Haag")
        expectedJSON.put("state", "CA")
        expectedJSON.put("postalCode", "2585 GJ")
        expectedJSON.put("countryCode", "NL")
        expectedJSON.put("experienceProfile", JSONObject().apply {
            put("noShipping", true)
            put("brandName", "My Brand!")
        })

        JSONAssert.assertEquals(expectedJSON, JSONObject(requestBody), true)
    }

    @Test
    fun createPaymentMethod_onPOSTError_forwardsErrorToCallback() {
        val error = Exception("error")
        val braintreeClient = MockkBraintreeClientBuilder()
            .sendPOSTErrorResponse(error)
            .build()

        val sut = LocalPaymentApi(braintreeClient, analyticsParamRepository, merchantRepository)

        sut.createPaymentMethod(
            getIdealLocalPaymentRequest(),
            localPaymentInternalAuthRequestCallback
        )

        verify { localPaymentInternalAuthRequestCallback.onLocalPaymentInternalAuthResult(null, error) }
    }

    @Test
    fun createPaymentMethod_onJSONError_forwardsJSONErrorToCallback() {
        val braintreeClient = MockkBraintreeClientBuilder()
            .sendPOSTSuccessfulResponse(Fixtures.ERROR_RESPONSE)
            .build()

        val sut = LocalPaymentApi(braintreeClient, analyticsParamRepository, merchantRepository)

        sut.createPaymentMethod(
            getIdealLocalPaymentRequest(),
            localPaymentInternalAuthRequestCallback
        )

        val captor = slot<Exception>()
        verify { localPaymentInternalAuthRequestCallback.onLocalPaymentInternalAuthResult(null, capture(captor)) }

        // Accept any Exception, or check for the actual type thrown by your implementation
        assertNotNull(captor.captured)
    }

    @Test
    fun createPaymentMethod_onPOSTSuccess_returnsResultWithOriginalRequestToCallback() {
        val braintreeClient = MockkBraintreeClientBuilder()
            .sendPOSTSuccessfulResponse(Fixtures.PAYMENT_METHODS_LOCAL_PAYMENT_CREATE_RESPONSE)
            .build()

        val sut = LocalPaymentApi(braintreeClient, analyticsParamRepository, merchantRepository)

        val request = getIdealLocalPaymentRequest()
        sut.createPaymentMethod(request, localPaymentInternalAuthRequestCallback)

        val captor = slot<LocalPaymentAuthRequestParams>()
        verify { localPaymentInternalAuthRequestCallback.onLocalPaymentInternalAuthResult(capture(captor), null) }

        val result = captor.captured
        assertNotNull(result)
        assertSame(request, result.request)
        assertEquals(
            "https://checkout.paypal.com/latinum?token=payment-token",
            result.approvalUrl
        )
        assertEquals("local-payment-id-123", result.paymentId)
    }

    @Test
    @Throws(JSONException::class)
    fun tokenize_sendsCorrectPostParams() {
        val braintreeClient = MockkBraintreeClientBuilder()
            .build()

        val sut = LocalPaymentApi(braintreeClient, analyticsParamRepository, merchantRepository)

        val webUrl = "sample-scheme://local-payment-success?paymentToken=successTokenId"
        sut.tokenize(
            "local-merchant-account-id", webUrl, "sample-correlation-id",
            localPaymentInternalTokenizeCallback
        )

        val captor = slot<String>()
        verify { braintreeClient.sendPOST(eq("/v1/payment_methods/paypal_accounts"), capture(captor), any(), any()) }
        val requestBody = captor.captured

        val expectedJSON = JSONObject()
        expectedJSON.put("merchant_account_id", "local-merchant-account-id")

        val paypalAccount = JSONObject()
            .put("intent", "sale")
            .put("response", JSONObject().put("webURL", webUrl))
            .put("options", JSONObject().put("validate", false))
            .put("response_type", "web")
            .put("correlation_id", "sample-correlation-id")
        expectedJSON.put("paypal_account", paypalAccount)

        val metaData = JSONObject()
            .put("source", "client")
            .put("integration", "custom")
            .put("sessionId", "sample-session-id")
        expectedJSON.put("_meta", metaData)

        JSONAssert.assertEquals(expectedJSON, JSONObject(requestBody), true)
    }

    @Test
    fun tokenize_onPOSTError_forwardsErrorToCallback() {
        val error = Exception("error")
        val braintreeClient = MockkBraintreeClientBuilder()
            .sendPOSTErrorResponse(error)
            .build()

        val sut = LocalPaymentApi(braintreeClient, analyticsParamRepository, merchantRepository)

        val webUrl = "sample-scheme://local-payment-success?paymentToken=successTokenId"
        sut.tokenize(
            "local-merchant-account-id", webUrl, "sample-correlation-id",
            localPaymentInternalTokenizeCallback
        )

        verify { localPaymentInternalTokenizeCallback.onResult(null, error) }
    }

    @Test
    fun tokenize_onJSONError_forwardsErrorToCallback() {
        val braintreeClient = MockkBraintreeClientBuilder()
            .sendPOSTSuccessfulResponse("not-json")
            .build()

        val sut = LocalPaymentApi(braintreeClient, analyticsParamRepository, merchantRepository)

        val webUrl = "sample-scheme://local-payment-success?paymentToken=successTokenId"
        sut.tokenize(
            "local-merchant-account-id", webUrl, "sample-correlation-id",
            localPaymentInternalTokenizeCallback
        )

        val captor = slot<Exception>()
        verify { localPaymentInternalTokenizeCallback.onResult(null, capture(captor)) }

        assertTrue(captor.captured is JSONException)
    }

    @Test
    fun tokenize_onPOSTSuccess_returnsResultToCallback() {
        val braintreeClient = MockkBraintreeClientBuilder()
            .sendPOSTSuccessfulResponse(Fixtures.PAYMENT_METHODS_LOCAL_PAYMENT_RESPONSE)
            .build()

        val sut = LocalPaymentApi(braintreeClient, analyticsParamRepository, merchantRepository)

        val webUrl = "sample-scheme://local-payment-success?paymentToken=successTokenId"
        sut.tokenize(
            "local-merchant-account-id", webUrl, "sample-correlation-id",
            localPaymentInternalTokenizeCallback
        )

        val captor = slot<LocalPaymentNonce>()
        verify { localPaymentInternalTokenizeCallback.onResult(capture(captor), null) }

        val result = captor.captured
        assertNotNull(result)
        assertEquals("e11c9c39-d6a4-0305-791d-bfe680ef2d5d", result.string)
        assertEquals("084afbf1db15445587d30bc120a23b09", result.clientMetadataId)
        assertEquals("jon@getbraintree.com", result.email)
        assertEquals("Jon", result.givenName)
        assertEquals("Doe", result.surname)
        assertEquals("9KQSUZTL7YZQ4", result.payerId)

        val shippingAddress = result.shippingAddress
        assertEquals("Jon Doe", shippingAddress.recipientName)
        assertEquals("836486 of 22321 Park Lake", shippingAddress.streetAddress)
    }

    private fun getIdealLocalPaymentRequest(): LocalPaymentRequest {
        val address = PostalAddress()
        address.streetAddress = "836486 of 22321 Park Lake"
        address.extendedAddress = "Apt 2"
        address.countryCodeAlpha2 = "NL"
        address.locality = "Den Haag"
        address.region = "CA"
        address.postalCode = "2585 GJ"

        val request = LocalPaymentRequest(true)
        request.paymentType = "ideal"
        request.amount = "1.10"
        request.address = address
        request.phone = "639847934"
        request.email = "jon@getbraintree.com"
        request.givenName = "Jon"
        request.surname = "Doe"
        request.isShippingAddressRequired = false
        request.merchantAccountId = "local-merchant-account-id"
        request.currencyCode = "EUR"
        request.paymentTypeCountryCode = "NL"
        request.displayName = "My Brand!"

        return request
    }
}